doc/sphinx/addendum/generalized-rewriting.rst

```diff
@@ -756,15 +756,70 @@ Subrelations are implemented in ``Classes.Morphisms`` and are a prime
 
 
-Constant unfolding
-~~~~~~~~~~~~~~~~~~
-
-The resolution tactic is based on typeclasses and hence regards user-defined
-:term:`constants <constant>` as transparent by default. This may slow down the
-resolution due to a lot of unifications (all the declared ``Proper``
-instances are tried at each node of the search tree). To speed it up,
-declare your constant as rigid for proof search using the command
-:cmd:`Typeclasses Opaque`.
+Constant unfolding during rewriting
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+  :n:`Hint {| Variables | Constants } {| Opaque | Transparent } : rewrite.`
+
+  :n:`Hint {| Opaque | Transparent } {+ @qualid } : rewrite.`
+
+   By default, :tacn:`setoid_rewrite` and :tacn:`rewrite_strat` unfold global
+   definitions during rewriterule matching, but do not unfold local
+   definitions. Unfolding definitions may slow down matching, whereas keeping
+   definitions opaque may cause matches to be missed.  This behavior is
+   configurable using transparency hints in the hint database ``rewrite`` as
+   shown, similar to the effect of :cmd:`Hint Transparent` on unification used
+   by :tacn:`eauto` and related tactics. Note that the transparency information
+   from database `rewrite` is used even when rewriting with individual lemmas.
+
+   .. coqtop:: reset in
+
+      Require Setoid.
+      Definition f x y := 2*x + y.
+      Definition g x y := 2*x + y.
+
+      Goal forall
+        (double_f : forall x y, 2*f x y = f (2*x) y + y),
+        2 * g 1 8 = 20.
+      Proof.
+
+   .. coqtop:: all
+
+        intros.
+          assert_succeeds (setoid_rewrite double_f).
+          assert_succeeds (rewrite_strat bottomup double_f).
+        set (x := g _).
+          assert_fails (setoid_rewrite double_f).
+          assert_fails (rewrite_strat bottomup double_f).
+        Hint Variables Transparent : rewrite.
+          assert_succeeds (setoid_rewrite double_f).
+          assert_succeeds (rewrite_strat bottomup double_f).
+        Hint Constants Opaque : rewrite.
+          assert_fails (setoid_rewrite double_f).
+          assert_fails (rewrite_strat bottomup double_f).
+        subst x.
+          assert_fails (setoid_rewrite double_f).
+          assert_fails (rewrite_strat bottomup double_f).
+        Hint Transparent f g : rewrite.
+          assert_succeeds (setoid_rewrite double_f).
+          assert_succeeds (rewrite_strat bottomup double_f).
+
+   .. coqtop:: none
+
+        exact eq_refl.
+      Qed.
+
+Constant unfolding during Proper-instance search
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+  :n:`Hint {| Opaque | Transparent } {+ @qualid } : typeclass_instances.`
+
+``Proper`` instances are resolved using typeclass search. By default, all
+constants are treated as transparent. This may slow down the resolution due to
+a lot of unifications (all the declared ``Proper`` instances are tried at each
+node of the search tree). To speed up the search, declare your non-abbreviation
```

Comment:
```suggestion
constants are treated as transparent. This may slow down the resolution because
:tacn:`typeclasses eauto` will do a lot of unifications (all the declared ``Proper``
instances are tried at each node of the proof search tree).
To speed up the search, declare your non-abbreviation
```